docs: improve data fetching section (#23420)
nuxt/nuxt: docs/1.getting-started/6.data-fetching.md

The first argument of [`useAsyncData`](/docs/api/composables/use-async-data) is the unique key used to cache the response of the second argument, the querying function. This argument can be ignored by directly passing the querying function. In that case, it will be auto-generated.

```ts
const { data, error } = await useAsyncData('users', () => myGetFunction('users'))
```

Since the autogenerated key only takes into account the file and line where `useAsyncData` is invoked, is recommended to always create your own key to avoid unwanted behavior.

```ts
const id = ref(1)

const { data, error } = await useAsyncData(`user:${id.value}`, () => {
  return myGetFunction('users', { id: id.value })
})
```

The `useAsyncData` composable is a great way to wrap and wait for multiple `useFetch` to be done, and then retrieve the results of each.

```ts
const { data: discounts, pending } = await useAsyncData('cart-discount', async () => {

>   /* Fetchs with no `await` will return promises */
>   const coupons = useFetch('cart/coupons')
>   const offers = useFetch('cart/offers')
> 
>   /* Wait for all fetches to complete */
>   Promise.all([coupons, offers])
> 
>   const 
> 
>   /* All set, return them */
>   return {
>     coupons: coupons.data,
>     offers: offers.data
>   }

Reviewer: ```suggestion
  const [coupons, offers] = await Promise.all([$fetch('/cart/coupons'), $fetch('/cart/offers')])

  return {
    coupons,
    offers
  }
```

})
```

::alert{type="warning"}

> Both `pick` and `transform` don't prevent the unwanted data to be transmitted. They only work over the data already received.

Reviewer: ```suggestion
Both `pick` and `transform` don't prevent the unwanted data from being fetched initially. But they will prevent unwanted data from being added to the payload transferred from server to client.
```

::

#### Watch

To re-run your fetching function each time other reactive values in your application change, use the `watch` option. You can use it for one or multiple _watchable_ elements.

```ts
const id = ref(1)

const { data, error, refresh } = await useFetch('/api/users', {
  /* Changing the id will trigger a refetch */
  watch: [id]
})
```

### Not immediate

The `useFetch` composable will start fetching data the moment is invoked. You may prevent this by setting `immediate: false`, for example, to wait for user interaction.

With that, you will need both the `status` to handle the fetch lifecycle, and `execute` to start the data fetch.

```vue
<script setup lang="ts">
const { data, error, execute, pending, status } = await useLazyFetch('/api/comments')
</script>

<template>
  <div v-if="status === 'idle'">
    <button @click="execute">Get data</button>
  </div>

  <div v-else-if="pending">
    Loading comments...
  </div>

  <div v-else>
    {{ data }}
  </div>
</template>
```

For finer control, the `status` variable can be:

> * `idle` when the fetch hasn't started,
> * `pending` when a fetch has started but not yet completed,
> * `error` when the fetch fails,
> * `success` when the fetch is completed successfully.

Reviewer: ```suggestion
* `idle` when the fetch hasn't started
* `pending` when a fetch has started but not yet completed
* `error` when the fetch fails
* `success` when the fetch is completed successfully
```
